src/condition.cpp

```diff
@@ -1274,7 +1274,7 @@ void conditional_t::set_query( const JsonObject &jo, const std::string &member,
     condition = [message, default_val, is_npc]( dialogue const & d ) {
         const talker *actor = d.actor( is_npc );
         if( actor->get_character() && actor->get_character()->is_avatar() ) {
-            std::string translated_message = _( message.evaluate( d ) );
+            std::string translated_message = to_translation( message.evaluate( d ) ).translated();
```

Comment:
This is equivalent to the previous line and not necessary, although I prefer `message.evaluate( d )` to be `translation` so it supports the translation context and text style checks.

Reply:
Cheers, I didn't realize it worked the same.

Tangent:
I just tried using a translation equivalent of `str_or_var` like you suggested, but it doesn't seem to pick up the translated text.

<details>
<summary>Tested with this diff</summary>

```diff
diff --git a/src/condition.cpp b/src/condition.cpp
index cfb2b275f0..0cb008af86 100644
--- a/src/condition.cpp
+++ b/src/condition.cpp
@@ -221,11 +221,14 @@ duration_or_var get_duration_or_var( const JsonObject &jo, const std::string &me
 }
 
 str_or_var get_str_or_var( const JsonValue &jv, const std::string &member, bool required,
-                           const std::string &default_val )
+                           const std::string &default_val, bool as_translation )
 {
     str_or_var ret_val;
     if( jv.test_string() ) {
         ret_val.str_val = jv.get_string();
+        if( as_translation && ret_val.str_val.has_value() ) {
+            ret_val.trans_str_val = to_translation( ret_val.str_val.value() );
+        }
     } else if( jv.test_object() ) {
         const JsonObject &jo = jv.get_object();
         if( jo.has_member( "mutator" ) ) {
@@ -239,6 +242,9 @@ str_or_var get_str_or_var( const JsonValue &jv, const std::string &member, bool
         jv.throw_error( "No valid value for " + member );
     } else {
         ret_val.str_val = default_val;
+        if( as_translation ) {
+            ret_val.trans_str_val = to_translation( default_val );
+        }
     }
     return ret_val;
 }
@@ -1269,12 +1275,12 @@ void conditional_t::set_one_in_chance( const JsonObject &jo, const std::string &
 
 void conditional_t::set_query( const JsonObject &jo, const std::string &member, bool is_npc )
 {
-    str_or_var message = get_str_or_var( jo.get_member( member ), member, true );
+    str_or_var message = get_str_or_var( jo.get_member( member ), member, true, "", true );
     bool default_val = jo.get_bool( "default" );
     condition = [message, default_val, is_npc]( dialogue const & d ) {
         const talker *actor = d.actor( is_npc );
         if( actor->get_character() && actor->get_character()->is_avatar() ) {
-            std::string translated_message = to_translation( message.evaluate( d ) ).translated();
+            std::string translated_message = message.evaluate( d );
             return query_yn( translated_message );
         } else {
             return default_val;
diff --git a/src/condition.h b/src/condition.h
index d3ec60f639..2ee85e4e93 100644
--- a/src/condition.h
+++ b/src/condition.h
@@ -56,7 +56,7 @@ const std::unordered_set<std::string> complex_conds = { {
 } // namespace dialogue_data
 
 str_or_var get_str_or_var( const JsonValue &jv, const std::string &member, bool required = true,
-                           const std::string &default_val = "" );
+                           const std::string &default_val = "", bool as_translation = false );
 dbl_or_var get_dbl_or_var( const JsonObject &jo, const std::string &member, bool required = true,
                            double default_val = 0.0 );
 dbl_or_var_part get_dbl_or_var_part( const JsonValue &jv, const std::string &member,
diff --git a/src/dialogue_helpers.cpp b/src/dialogue_helpers.cpp
index a5e189aae5..daf79c0841 100644
--- a/src/dialogue_helpers.cpp
+++ b/src/dialogue_helpers.cpp
@@ -79,6 +79,9 @@ std::string str_or_var::evaluate( dialogue const &d ) const
     if( function.has_value() ) {
         return function.value()( d );
     }
+    if( !trans_str_val.empty() ) {
+        return trans_str_val.translated();
+    }
     if( str_val.has_value() ) {
         return str_val.value();
     }
diff --git a/src/dialogue_helpers.h b/src/dialogue_helpers.h
index 929b9cb1b5..12f505e759 100644
--- a/src/dialogue_helpers.h
+++ b/src/dialogue_helpers.h
@@ -8,6 +8,7 @@
 #include "global_vars.h"
 #include "math_parser.h"
 #include "rng.h"
+#include "translation.h"
 #include "type_id.h"
 
 struct dialogue;
@@ -150,6 +151,7 @@ var_info process_variable( const std::string &type );
 
 struct str_or_var {
     std::optional<std::string> str_val;
+    translation trans_str_val;
     std::optional<var_info> var_val;
     std::optional<std::string> default_val;
     std::optional<std::function<std::string( const dialogue & )>> function;
```

</details>

But if I wrap `message.evaluate( d )` with `_( )` or `to_translation( ).translated()`, it suddenly translates correctly again. `message` gets a translation object with the english text, but it looks like it's loaded before translation strings? And recreating the translation when the condition gets called seems to get it cached properly?

Reply:
What I had in mind is to make a separate class (`translation_or_var`) that reads in `translation` directly using `JsonValue::read()` because using `JsonValue::get_string()` will only read a string without checking for the text style and cannot read an object that contains the translation context.

I cannot immediately see why it does not translate the string correctly in your code though, `translation` objects only stores the raw string so calling `transation::translated` should return the translated text when called right before the `query_yn` call.

Reply:
Oh, it looks like `read()` was the key there. Followed your advice using a dedicated `translation_or_var` object, and everything seems to be working :)